### ISSUE TITLE
RFE 1.3 : Better client connection management from (X) clients each performing (Y) ops.

### ISSUE THREAD

**Issue by timothysc** [kubernetes/kubernetes#21081]
During some of our testing on openshift, we had performed several workflows which did the following: 

1. Forked a new process with a new kubeconfig 
2. Performed several commands using kubectl (~10-20) 
3. Gather forks. 
4. repeat. 

What this showed was that the sum of connection management and tls overhead was quite large.  
Ideally we could reduce the overhead by reusing the connection which is done in many of the e2e tests.   

Some ideas here include: 
 - The notion of a repl, or shell, for kubectl 
 - Several different tools for client side mgmt
 - ...  

/cc @jeremyeder @kubernetes/sig-scalability @smarterclayton @liggitt

**Comment by jeremyeder** [comment]
In Linux, socket setup/teardown is an expensive operation in terms of CPU cycles and time.

Add TLS and it's more expensive.
Add SELinux (netlabels) and it's more expensive.  The most expensive combination possible :-1: 

We know that efficient communications are batched, pipelined and guided towards throughput over latency.  The amount of tcp_socket events we're seeing during load tests is surprisingly high.

This leads to higher than necessary CPU utilization.  We think improving the efficiency of communications will be key to Kube scaling.  Just wanted to spell that out, having worked on optimizing socket setup and teardown efficiency several years ago...

Here's the top few lines of pprof data from @timothysc scenario taken from the master node:

```
4651.78s of 6475.52s total (71.84%)
Dropped 3711 nodes (cum <= 32.38s)
      flat  flat%   sum%        cum   cum%
     0.02s 0.00031% 0.00031%   5383.28s 83.13%  runtime.goexit
     0.26s 0.004% 0.0043%   2592.50s 40.04%  net/http.(*conn).serve
     0.95s 0.015% 0.019%   1854.96s 28.65%  crypto/tls.(*Conn).Handshake
     0.08s 0.0012%  0.02%   1516.52s 23.42%  crypto/tls.(*Conn).serverHandshake
     0.36s 0.0056% 0.026%   1502.17s 23.20%  crypto/tls.(*serverHandshakeState).doFullHandshake
     0.20s 0.0031% 0.029%   1448.41s 22.37%  crypto/rsa.SignPKCS1v15
     0.21s 0.0032% 0.032%   1447.47s 22.35%  crypto/rsa.decrypt
     0.71s 0.011% 0.043%   1330.91s 20.55%  math/big.nat.expNN
     0.18s 0.0028% 0.046%   1328.69s 20.52%  math/big.(*Int).Exp
     9.60s  0.15%  0.19%   1254.50s 19.37%  math/big.nat.expNNWindowed
     0.31s 0.0048%   0.2%   1222.57s 18.88%  crypto/tls.(*ecdheKeyAgreement).generateServerKeyExchange
     0.52s 0.008%  0.21%   1114.91s 17.22%  net/http.HandlerFunc.ServeHTTP
     0.13s 0.002%  0.21%   1101.33s 17.01%  net/http.(*ServeMux).ServeHTTP
     0.09s 0.0014%  0.21%   1100.53s 17.00%  github.com/emicklei/go-restful.Container.ServeHTTP
     0.11s 0.0017%  0.21%   1099.80s 16.98%  github.com/emicklei/go-restful.(*Container).ServeHTTP
````

**Comment by smarterclayton** [comment]
Go elliptic curve crypto is also notoriously slow, and improvements are
blocked on someone contributing a non-license encumbered implementation.
 20x or 30x as slow as it could be on a modern implementation.

**Comment by timothysc** [comment]
Optimizing the tls handshake is important, but as multi-tenancy grows... so grows connection mgmt which is also important.

**Comment by jeremyeder** [comment]
@smarterclayton aka never?  Seriously...  The encryption overhead is proportional to the # of connections.  Improve connection efficiency and encryption overhead also goes down.  Two birds!

**Comment by smarterclayton** [comment]
I'm not sure this is an important overall scenario to optimize - since kubectl interactions are expected to be a small % of our overall API workload (I would expect <10% on an average cluster, maybe <2%).

I agree that we are wasting CPU cycles on command-line clients connecting to the cluster.  The story now is to drop to non-elliptic crypto or put a TLS terminator in front that uses an OpenSSL backed crypto implementation that leverages AES NI.  I don't think we expect movement on golang crypto in this area even in 1.7 (or haven't seen anything recently to lead me to believe it would happen).

**Comment by timothysc** [comment]
On the node+controller client connections come & go too often vs. pooling: 

![image](https://cloud.githubusercontent.com/assets/169553/13019984/5e9895d6-d198-11e5-9a23-f3e9cb32ce01.png)

In the graph above is 100 node cluster, and before anything is running we have 2000 connections.   The lumps in the graph are the ```kubectl``` events, but you notice they decrease as the clients terminate.  But the overall trend in the system steadily increases with load.

**Comment by smarterclayton** [comment]
If you're talking about cluster to cluster connections, yes, I would expect there to be a very, very, very low level of new connection establishment.  This sounds like something broke, but it's not really kubectl related (that's a disjoint problem).

**Comment by smarterclayton** [comment]
The last time this was watch reestablishment.  It's also possible to exceed the pool size - I believe we don't have tunables for the etcd transport or the controller -> master pool size either, so we should definitely ensure those are properly set.  At most I would expect 1-2 connections established per controller loop per minute, and that should be tunable

**Comment by smarterclayton** [comment]
It's also possible that the controller parallelism got tuned up and as a result we're getting a lot of wasted spin up - the rate limiter used to smooth that out (as long as the initial connection time < the rate).  If we upped the rate limiter we would then have a higher wasted connection rate.

**Comment by ncdc** [comment]
cc @kubernetes/rh-cluster-infra

**Comment by smarterclayton** [comment]
We really, really, really, really need a way to assign each connection
creation to a caller trace.

**Comment by ncdc** [comment]
I started fiddling with this a bit. If you add conn tracking to `net.TCPConn`, with a simple test of calling `http.Get()`, you don't get useful info because the TCPConn's are created in a goroutine from `net/http.(*Transport).getConn()`, so you don't see the actual top-level function that initiated the connection. I'll keep playing with this - I might just need to move the tracking to multiple places like getConn.

**Comment by timothysc** [comment]
@rrati I believe there was metrics hacks that allowed us to track per-host connection counts+endpoints on the master.  

/cc @jayunit100

**Comment by timothysc** [comment]
During our @kubernetes/sig-scalability we had discussed whether golang 1.6 w/ http2 might be a clean way of dealing with this.  As http2 supports connection multiplexing.

**Comment by timothysc** [comment]
We've got the new cluster up with latest and greatest + golang 1.6.  Initial findings show HTTP2 is enabled on the server by default w/golang 1.6, but the clients still showing many connections, and am digging in now.

**Comment by liggitt** [comment]
go1.6 should have made the per-handshake TLS overhead better, but the "connection per CLI invocation" wouldn't have changed

**Comment by timothysc** [comment]
Closing, fixed via #25280
